### PR TITLE
Make docs action more reusable

### PR DESCRIPTION
**Pull request by benjeffery**
Until github implement https://github.com/actions/runner/issues/646 we're stuck copy pasting the actions code, I've made it more copy friendly here.

Don't merge yet, as need PR for testing.

### CODE REVIEW

**Comment by AdminBot-tskit** [comment]
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/msprime-docs/1281/)

**Comment by AdminBot-tskit** [comment]
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/msprime-docs/PR-1281/)

**Comment by benjeffery** [comment]
@jeromekelleher This is ready to merge. I've had to make the docs deterministic otherwise we get a doc preview comment on every PR. This will need a manual merge.

File: docs/ancestry.rst
Comment (jeromekelleher) [inline docs/ancestry.rst:957]: Something strange has happened here - did you pick up some extra stuff in a rebase? This PR probably shouldn't touch this file.

**Review by jeromekelleher — COMMENTED** [review]
Looks good, but see query about the changes to docs.

File: docs/ancestry.rst
Comment (benjeffery) [reply docs/ancestry.rst:957]: This section of the docs gave different HTML every time the docs were built. I know it isn't ideal, but this was the only way I could see to get the point across about randomness while using deterministic HTML.

File: .github/workflows/docs.yml
Comment (grahamgower) [inline .github/workflows/docs.yml:99]: This bit is very clever. But I just want to get some clarification of the behaviour here... If we merge PR `#aaa`, this action will run with the `main` branch trigger (or whatever it's called), deleting the docs for `#aaa`. But if we close PR `#aaa` without merging, this won't run? If someone opens PR `#bbb` after `#aaa` is closed-without-merging, then the `#aaa` docs are deleted, right? Is there also a race condition in here where two instances of the admin bot push commits with the same docs folder deleted?

Comment (benjeffery) [reply .github/workflows/docs.yml:99]: You've got that right - I thought it was probably OK for a closed PR's docs to hang around until the next run of the action.

As for race conditions - if two actions are simultaneously running then who ever got the push in last would wipe out the first actions commit. To mitigate this I've added the `turnstyle` step which waits for other doc actions to finish.

Comment (grahamgower) [reply .github/workflows/docs.yml:99]: Got it! Thanks!

Comment (benjeffery) [reply .github/workflows/docs.yml:99]: :+1: Thanks for looking over these, this stuff is hard to test so a second pair of eyes is great.

File: docs/ancestry.rst
Comment (jeromekelleher) [reply docs/ancestry.rst:957]: I see, sounds good.

**Review by jeromekelleher — APPROVED** [review]

**Comment by AdminBot-tskit** [comment]
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/msprime-docs/PR-/)

**Pull request merged**
